storage/src/checker/mod.rs

```diff
@@ -42,19 +43,34 @@ impl<S: WritableStorage> NodeStore<Committed, S> {
         let mut visited = LinearAddressRangeSet::new(db_size)?;
 
         // 2. traverse the trie and check the nodes
-        if let (Some(root_address), Some(root_hash)) = (self.root_address(), self.root_hash()) {
+        if let Some(root) = self.root_as_maybe_persisted_node() {
             // the database is not empty, traverse the trie
-            self.check_area_aligned(
-                root_address,
-                StoredAreaParent::TrieNode(TrieNodeParent::Root),
-            )?;
-            self.visit_trie(
-                root_address,
-                HashType::from(root_hash),
-                Path::new(),
-                &mut visited,
-                opt.hash_check,
-            )?;
+            if let Some(root_address) = root.as_linear_address() {
+                self.check_area_aligned(
+                    root_address,
+                    StoredAreaParent::TrieNode(TrieNodeParent::Root),
+                )?;
+                // For hash checking, we need the root hash if it exists
+                let root_hash = self.root_hash().map(|h| h.into_hash_type());
+                if let Some(root_hash) = root_hash {
+                    self.visit_trie(
+                        root_address,
+                        root_hash,
+                        Path::new(),
+                        &mut visited,
+                        opt.hash_check,
+                    )?;
+                } else {
+                    // No root hash available, visit without hash checking
```

Comment:
Why would there ever be a trie without root hash? I think it is fine to compute the root hash before traversing for completeness, but we can also change this parameter to `Option<HashType>` so that we still do hash checks on children nodes.